File: 79.word-search/solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    bool existRec(vector<vector<char>> &board, string word, int i, int j)
    {
        if (word.size() == 0)
        {
            return true;
        }

        if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size())
        {
            return false;
        }

        if (board[i][j] == '#')
        {
            return false;
        }

        char temp = board[i][j];
        if (board[i][j] == word[0])
        {
            board[i][j] = '#';
            bool hasSolution = this->existRec(board, word.substr(1), i + 1, j) ||
                               this->existRec(board, word.substr(1), i - 1, j) ||
                               this->existRec(board, word.substr(1), i, j + 1) ||
                               this->existRec(board, word.substr(1), i, j - 1);
            
            if (hasSolution) {
                return true;
            }
        }

        board[i][j] = temp;
        return false;
    }
    bool exist(vector<vector<char>> &board, string word)
    {
        int m = board.size();
        int n = board[0].size();

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (board[i][j] == word[0] && this->existRec(board, word, i, j))
                {
                    return true;
                }
            }
        }
        return false;
    }
};

int main(int argc, char const *argv[])
{

    return 0;
}

File: 79.word-search/solution.py
from typing import List


class Solution:
    def existsRec(self, board: List[List[str]], word: str, i: int, j: int, visited: List[List[bool]]) -> bool:
        if len(word) == 0:
            return True

        if i >= len(board) or j >= len(board[0]) or i < 0 or j < 0:
            return False

        if visited[i][j]:
            return False

        if board[i][j] == word[0]:
            visited[i][j] = True
            if (self.existsRec(board, word[1:], i - 1, j, visited)
                or self.existsRec(board, word[1:], i + 1, j, visited)
                or self.existsRec(board, word[1:], i, j + 1, visited)
                    or self.existsRec(board, word[1:], i, j - 1, visited)):
                return True
            else:
                visited[i][j] = False

        return False

    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])

        explored_board = [[False for _ in range(n)] for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0] and self.existsRec(board, word, i, j, explored_board):
                    return True

        return False


board = [["A", "A", "A", "A", "A", "A"], ["A", "A", "A", "A", "A", "A"], ["A", "A", "A", "A", "A", "A"], [
    "A", "A", "A", "A", "A", "A"], ["A", "A", "A", "A", "A", "A"], ["A", "A", "A", "A", "A", "A"]]
word = "AAAAAAAAAAAAAAB"

sol = Solution()
print(sol.exist(board, word))
